homeassistant/components/denonavr/config_flow.py

```diff
@@ -0,0 +1,239 @@
+                return await self.async_step_settings()
+
+            # discovery using denonavr library
+            self.d_receivers = await self.hass.async_add_executor_job(denonavr.discover)
+            # More than one receiver could be discovered by that method
+            if len(self.d_receivers) == 1:
+                self.host = self.d_receivers[0]["host"]
+                return await self.async_step_settings()
+            if len(self.d_receivers) > 1:
+                # show selection form
+                return await self.async_step_select()
+
+            errors["base"] = "discovery_error"
+
+        return self.async_show_form(
+            step_id="user", data_schema=CONFIG_SCHEMA, errors=errors
+        )
+
+    async def async_step_select(self, user_input=None):
+        """Handle multiple receivers found."""
+        errors = {}
+        if user_input is not None:
+            self.host = user_input["select_host"]
+            return await self.async_step_settings()
+
+        select_scheme = vol.Schema(
+            {
+                vol.Required("select_host"): vol.In(
+                    [d_receiver["host"] for d_receiver in self.d_receivers]
+                )
+            }
+        )
+
+        return self.async_show_form(
+            step_id="select", data_schema=select_scheme, errors=errors
+        )
+
+    async def async_step_settings(self, user_input=None):
+        """Allow the user to specify settings."""
+        if user_input is not None:
+            # Get config option that have defaults
+            self.show_all_sources = user_input[CONF_SHOW_ALL_SOURCES]
+            self.zone2 = user_input[CONF_ZONE2]
+            self.zone3 = user_input[CONF_ZONE3]
+
+            return await self.async_step_connect()
+
+        return self.async_show_form(step_id="settings", data_schema=SETTINGS_SCHEMA)
+
+    async def async_step_connect(self, user_input=None):
+        """Connect to the receiver."""
+        connect_denonavr = ConnectDenonAVR(
+            self.hass,
+            self.host,
+            self.timeout,
+            self.show_all_sources,
+            self.zone2,
+            self.zone3,
+        )
+        if not await connect_denonavr.async_connect_receiver():
+            return self.async_abort(reason="connection_error")
+        receiver = connect_denonavr.receiver
+
+        mac_address = await self.async_get_mac(self.host)
+
+        if not self.serial_number:
+            self.serial_number = receiver.serial_number
+        if not self.model_name:
+            self.model_name = (receiver.model_name).replace("*", "")
+
+        if not self.serial_number:
+            _LOGGER.error(
+                "Could not get serial number of host %s, "
+                "using the mac_address as identification",
+                self.host,
+            )
+            if not mac_address:
+                return self.async_abort(reason="no_mac")
+            self.serial_number = mac_address
```

Comment:
We shouldn't set the mac address as the serial number and vice versa. Please keep these variables separate.

Introduce a third variable if needed to hold the unique_id that should be used for setting the unique_id.

Reply:
I made a PR that prevents mixing of the serial_number as mac_address.

The other way around is more difficult since the mac_adress is also needed downsteam in the __init__.py file and again in the media_player.py file to idintify the enitity with the device.

Otherwise I would need to add an extra paramter to the Config entry, something like "device_identification" that will be identical to the mac_adress in normal cases, but will be diffrent if getting the mac failed.

However I think that is also messy since you then have another config flow data variable.
@MartinHjelmare Schould I add that extra var or leave it like this?

Reply:
We should definitely not store the serial number as a mac address in the config entry data. Please change that. Name the data correctly and don't use the serial number for things that should use a mac address, like the connection set in device_info.

Reply:
done no mixing anymore